### ISSUE TITLE
Undefined method on using Puma on Heroku

### ISSUE THREAD

**Issue by daqo** [puma/puma#651]
My RoR application works smoothly when deployed using Webrick on Heroku. I tried to add puma (and at a later try, unicorn) as the default web server, everything goes well but one thing.

I'm using RGeo gem to keep track of some hive locations. When switched to Puma, calling ``@point.current_location.lon`` and ``@point.current_location.lat`` throws the error:

``ActionView::Template::Error (undefined method 'lon' and 'lat' for #<RGeo::Cartesian::PointImpl:0x007f700846c970>)``

Meanwhile, I tried to access the current_location latitude and longitude through heroku run rails console

The interesting thing is that fetching lat and long values on Heroku console returns correct values. It seems that the problem results from puma and unicorn.

Any suggestion?

**Comment by evanphx** [comment]
I can only guess there is some thread unsafe code in that path that is causing this. When the code works, what is the class? Is it  still that `RGeo::Cartesian::PointImpl`? Are `lon` and `lat` normal methods on it? I can't see a reason the methods would disappear unless there is some metaprogramming there that is run on first use (and thus causes threading issues).

**Comment by daqo** [comment]
Yes, the class is ``PointImpl``. On my local production server (Not heroku) I'm using Passenger for production. Over there, the app works fine. If the problem was a thread unsafe code, shouldn't I have the same problem over there as well?

**Comment by evanphx** [comment]
Well, I'm not sure how puma would influence this code other than running it in threads. As for why it into happens in puma, threading issues are so finicky it could be any number of factors like the number of threads involved.

**Comment by daqo** [comment]
Right now I checked it with Unicorn too. (On Heroku)
It seems that Unicorn has the same problem with that piece of code. So I think the problem goes back to the way Heroku manages stuff; though, I'm not sure.

**Comment by evanphx** [comment]
Does this happen with every request too?

**Comment by daqo** [comment]
Yes, every single request. Lat and Lon method will get disappeared.

**Comment by evanphx** [comment]
Ah! Then yeah, it's not threading. The RGeo library isn't loading properly I guess.

**Comment by evanphx** [comment]
So I don't see the `lat` or `lon` methods in the RGeo repo even...

**Comment by evanphx** [comment]
Ok, I see it now. It's a method on `SphericalPointImpl` though, not `PointImpl`. How are you creating the object? Perhaps there is a factory or something that is supposed to be making a `SphericalPointImpl` but is returning a `PointImpl` instead?

**Comment by daqo** [comment]
So basically, in my model I use:
 ```
  set_rgeo_factory_for_column(:current_location, 
                              RGeo::Geographic.spherical_factory(:srid => 4326))
```

And then in my location creator I have the following code:

```
require_relative "./coordinate_builder"
require 'rgeo'

class LocationCreator
  DEFAULT = "default"

  def self.create(lat_lng)
    return nil if lat_lng == DEFAULT

    lng, lat = CoordinateBuilder.parse(lat_lng)
    geographic_factory = RGeo::Geographic.spherical_factory
    geographic_factory.point(lng, lat)
  end
end
```

**Comment by daqo** [comment]
Ok I found the problem Evan.

It had nothing to do with Puma. Apparently [Rgeo::Geographic::SphericalPointImpl falls back to RGeo::Cartesian::PointImpl](https://groups.google.com/forum/#!topic/rgeo-users/mFyFWov5UoU) in my Javascript code. As a workaround, I changed my javascript calls from ```@hive.current_location.lat``` to ```@hive.current_location.y```. Also for latitude, I did the same: ```@hive.current_location.lon``` to ```@hive.current_location.x```. This fixed the problem for me.

Thank you so much for your help.
